tokio/src/sync/mpsc/list.rs

```diff
@@ -230,6 +239,24 @@ impl<T> fmt::Debug for Tx<T> {
 }
 
 impl<T> Rx<T> {
+    pub(crate) fn is_empty(&self, tx: &Tx<T>) -> bool {
+        let block = unsafe { self.head.as_ref() };
+        if block.has_value(self.index) {
+            return false;
+        }
+
+        // It is possible that a block has no value "now" but the list is still not empty.
+        // To be sure, it is necessary to check the length of the list.
+        self.len(tx) == 0
+    }
+
+    pub(crate) fn len(&self, tx: &Tx<T>) -> usize {
+        // When all the senders are dropped, there will be a last block in the tail position,
+        // but it will be closed
+        let tail_position = tx.tail_position.load(Acquire);
+        tail_position - self.index - (tx.is_closed() as usize)
+    }
```

Comment:
Do you have a test for whether the length is correct in all of the following cases?

* There are still senders, but `Receiver::close` has been called.
* There are no more senders, but `Receiver::close` has not been called.
* There are no more senders, and `Receiver::close` has been called before the last sender dropped.
* There are no more senders, and `Receiver::close` has been called after the last sender dropped.

Reply:
Some of the scenarios were missing, but I have added them now.